src/cwt/claim.rs

```diff
@@ -0,0 +1,352 @@
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("claim requires String value")]
+    StringValueRequired,
+    #[error("claim requires NumericDate (int or float)")]
+    NumericDateRequired,
+    #[error("claim requires Null value")]
+    NullValueRequired,
+    #[error("claim requires Bool value")]
+    BoolValueRequired,
+    #[error("claim requires Integer value")]
+    IntegerValueRequired,
+    #[error("claim requires Float value")]
+    FloatValueRequired,
+    #[error("claim requires Bytes value")]
+    BytesValueRequired,
+    #[error("claim requires Text value")]
+    TextValueRequired,
+    #[error("claim requires Array value")]
+    ArrayValueRequired,
+    #[error("claim requires Map value")]
+    MapValueRequired,
+    #[error("claim requires Tag value")]
+    TagValueRequired,
+    #[error("generic parse error {0}")]
+    GenericParseError(String),
+    #[error("TryFromIntError: {0}")]
+    TryFromIntError(#[from] std::num::TryFromIntError),
+    #[error("OffsetComponentError: {0}")]
+    OffsetComponentError(#[from] time::error::ComponentRange),
+    #[error("NumericDate conversion overflows")]
+    ConversionOverflow,
+}
+
+/// Numerical representation of seconds relative to the Unix Epoch,
+/// as defined in [RFC7049](https://www.rfc-editor.org/rfc/rfc7049#section-2.4.1)
+/// with the leading tag 1 omitted.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum NumericDate {
+    IntegerSeconds(i128),
+    FractionalSeconds(f64),
+}
+impl From<NumericDate> for Value {
+    fn from(value: NumericDate) -> Self {
+        match value {
+            NumericDate::IntegerSeconds(i) => Value::Integer(i),
+            NumericDate::FractionalSeconds(f) => Value::Float(f),
+        }
+    }
+}
+
+const NANOS_PER_SECOND: u32 = 1_000_000_000;
+
+impl TryFrom<NumericDate> for OffsetDateTime {
+    type Error = Error;
+
+    /// For IntegerSeconds: Attempts to downcast from i128 -> 164,
+    /// before converting to UNIX timestamp.
+    /// For FractionalSeconds: truncates to (max) 9 decimal places.
+    fn try_from(value: NumericDate) -> Result<Self, Self::Error> {
+        match value {
+            NumericDate::IntegerSeconds(i) => {
+                let i: i64 = i.try_into()?;
+                Ok(OffsetDateTime::from_unix_timestamp(i)?)
+            }
+            NumericDate::FractionalSeconds(f) => {
+                println!("f: {:?}", f);
+                let nanos = (f * (NANOS_PER_SECOND as f64)).trunc();
+                println!("nanos: {:?}", nanos);
+                // OffsetDateTime is more restrictive,
+                // but defer to their error-handling where possible
+                if nanos >= (std::i128::MAX as f64) || nanos <= (std::i128::MIN as f64) {
+                    return Err(Error::ConversionOverflow);
+                }
+                let nanos: i128 = nanos as i128;
+                Ok(OffsetDateTime::from_unix_timestamp_nanos(nanos)?)
```

Comment:
I'm surprised there's no loss of precision here when we multiply the `f64` by `NANOS_PER_SECOND`, since you will likely end up with a number larger than 2^53. Do you know why we don't lose precision? If not could I suggest building the OffsetDateTime from the integer and fractional parts of the `f64` separately (build a OffsetDateTime with from UNIX epoch seconds, and then replace the nanoseconds)?

Reply:
From what I could tell investigating yesterday, the floats get represented in scientific notation and so don't really lose significant figures (beyond the ~15 that they can track) when multiplying by 10e9. That said, I think this could still be the source of the epsilon error in the ignored test case which is what I was continuing to look into, will try to go your route!

Reply:
Hmm yeah so I'm seeing roughly the same behavior when I switch to doing time + replacing nanoseconds. I think the issue is the same as before, which is that for a couple of cases (including something like the max timestamp with full nanos), the timestamp + nanoseconds (to 9 places) is too much over the 15ish sig figs available; after 15ish sig figs, the remaining nanoseconds are still just floating point error either way. I can look into trying to only construct an offset with up to 15 significant figures instead?